### PR TITLE
feat: replace oauth and userpoolclient lambdas with cfn and sdk calls

### PR DESCRIPTION
**Pull request by sobolk**
#### Description of changes

This PR:
- Incorporates changes from https://github.com/aws-amplify/amplify-cli/commits/feat/replace-lambda-callouts that replace oauth and userpoolclient custom CFN resource lambdas
- Makes changes to make existing e2e pass and adds couple of them

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-e2e-core/src/categories/storage.ts [hunk 1]
```diff
@@ -610,9 +610,9 @@ export function updateS3AddTriggerNewFunctionWithFunctionExisting(cwd: string, s
   });
 }
 
-export function addS3StorageWithIdpAuth(projectDir: string): Promise<void> {
+export function addS3StorageWithIdpAuth(projectDir: string, testingWithLatestCodebase = false): Promise<void> {
```
Comment (Amplifiyer) [inline packages/amplify-e2e-core/src/categories/storage.ts:613]: Shouldn't the default be true?

File: packages/amplify-e2e-tests/src/__tests__/auth/hosted-ui.test.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  initJSProjectWithProfile,
+  updateAuthSignInSignOutUrlWithAll,
+} from '@aws-amplify/amplify-e2e-core';
+
+const defaultsSettings = {
+  name: 'authTest',
+};
+
+const oauthSettings = {
+  signinUrl: 'https://danielle.lol/',
```
Comment (Amplifiyer) [inline packages/amplify-e2e-tests/src/__tests__/auth/hosted-ui.test.ts:19]: 🤣

**Review by Amplifiyer — DISMISSED** [review]

File: packages/amplify-e2e-core/src/categories/storage.ts (hunk 1 above)
Comment (sobolk) [reply packages/amplify-e2e-core/src/categories/storage.ts:613]: nah. The patter is "false", see https://github.com/search?q=repo%3Aaws-amplify%2Famplify-cli%20testingWithLatestCodebase&type=code . and then we rely on cci/codebuild to set this up properly ...

**Review by danielleadams — CHANGES_REQUESTED** [review]
Looks good. I think there's 1 change we need to make, and then I will say it looks like there's future tests that were added, but I think that's okay because if the tests pass now, we can just TDD the future code.

**Comment by sobolk** [comment]
> Looks good. I think there's 1 change we need to make, and then I will say it looks like there's future tests that were added, but I think that's okay because if the tests pass now, we can just TDD the future code.

Yes. Taking these tests eagerly was intentional.

File: codebuild_specs/e2e_workflow_generated.yml [hunk 3]
```diff
@@ -2158,7 +2166,7 @@ batch:
         variables:
           compute-type: BUILD_GENERAL1_SMALL
           TEST_SUITE: src/__tests__/migration_tests_v10/custom-stack.migration.test.ts
-          CLI_REGION: us-west-2
+          CLI_REGION: us-east-2
```
[the remark this answers is not in the document]
Comment (sobolk) [reply codebuild_specs/e2e_workflow_generated.yml:2169]: This was outcome from `yarn split-e2e-tests-codebuild`. We're supposed to run in and commit outcome if test suite changes.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 4]
```diff
@@ -68,12 +62,8 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
   identityPoolRoleMap: cognito.CfnIdentityPoolRoleAttachment | undefined;
   lambdaConfigPermissions?: Record<string, lambda.CfnPermission>;
   lambdaTriggerPermissions?: Record<string, iam.CfnPolicy>;
-  // custom resources userPoolClient
-  userPoolClientLambda?: lambda.CfnFunction;
+  // provides base role for deleting custom resource lambdas that are no longer needed after the migration
   userPoolClientRole?: iam.CfnRole;
-  userPoolClientLambdaPolicy?: iam.CfnPolicy;
```
[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:74]: We can't it's on customer facing type in overrides, we'd have to ship major version bump.
See https://github.com/aws-amplify/amplify-cli/blob/a3b139fb36bcb2c1c890403e5bd754beec5b7ff7/packages/amplify-cli-extensibility-helper/src/types/auth/types.ts#L33C11-L33C11 .

[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:74]: What we could do is deprecate old field and add new. But I'm not sure it's worth it.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 5]
```diff
@@ -837,83 +787,9 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
...
-        oAuthMetadata: cdk.Fn.ref('oAuthMetadata'),
-        webClientId: cdk.Fn.ref('UserPoolClientWeb'),
-        nativeClientId: cdk.Fn.ref('UserPoolClient'),
-        userPoolId: cdk.Fn.ref('UserPool'),
-      },
-    });
-    this.oAuthCustomResourceInputs.node.addDependency(this.oAuthCustomResourceLogPolicy);
+    // this can be removed when hostedUI Custom resource is removed
+    this.userPoolClient?.node.addDependency(this.hostedUIProvidersCustomResourceInputs);
+    this.userPoolClientWeb?.node.addDependency(this.hostedUIProvidersCustomResourceInputs);
```
[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts]: Changed in [74cf0dc](https://github.com/aws-amplify/amplify-cli/pull/12935/commits/74cf0dc89533de82ac269292cddbc6e39c670354) .
When I re-recorded snapshots, these e2e tests, so confidence is high. running full e2e on the change.

[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts]: I have to revert it. it caused failures across the board
image
https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/22063/workflows/8803c18b-c076-4eb1-9a7c-ce7b6901a3dd

[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:792]:
```
CREATE_COMPLETE                Thu Jul 13 2023 \u001b[39m\r\n\u001b[K\u001b[32m\tHostedUICustomResourceInputs   Custom::LambdaCallout          CREATE_COMPLETE                Thu Jul 13 2023 \u001b[39m\r\n\u001b[?25h\r\n"]
[214.859,"o","🛑 \u001b[31mThe following resources failed to deploy:\u001b[39m\r\n"]
[214.859,"o","\u001b[31mResource Name:\u001b[39m UserPoolClient (AWS::Cognito::UserPoolClient)\r\n\u001b[31mEvent Type:\u001b[39m create\r\n\u001b[31mReason:\u001b[39m The provider Facebook does not exist for User Pool us-east-2_OVbZj6wvz. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: f28ddb69-9b44-484e-87a8-ba40bb5c3cf6; Proxy: null)\r\n\r\n\r\n\u001b[31mResource Name:\u001b[39m UserPoolClientWeb (AWS::Cognito::UserPoolClient)\r\n\u001b[31mEvent Type:\u001b[39m create\r\n\u001b[31mReason:\u001b[39m The provider Facebook does not exist for User Pool us-east-2_OVbZj6wvz. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: 74dfe154-1c80-4e5f-b050-e20c70633db7; Proxy: null)\r\n\r\n\r\n"]
[214.86,"o","🛑 \u001b[31mResource is not in the state stackUpdateComplete\u001b[39m\r\n"]
[214.86,"o","\u001b[0mName: UserPoolClient (AWS::Cognito::UserPoolClient), Event Type: create, Reason: The provider Facebook does not exist for User Pool us-east-2_OVbZj6wvz. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: f28ddb69-9b44-484e-87a8-ba40bb5c3cf6; Proxy: null), IsCustomResource: false\u001b[0m\r\n\u001b[0m\u001b[0m\r\n\u001b[0mName: UserPoolClientWeb (AWS::Cognito::UserPoolClient), Event Type: create, Reason: The provider Facebook does not exist for User Pool us-east-2_OVbZj6wvz. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: 74dfe154-1c80-4e5f-b050-e20c70633db7; Proxy: null), IsCustomResource: false\u001b[0m\r\n\u001b[0m\u001b[0m\r\n\r\n"]
[214.86,"o","\u001b[0mLearn more at: https://docs.amplify.aws/cli/project/troubleshooting/\u001b[0m\r\n\r\nDeploymentFault: Resource is not in the state stackUpdateComplete\r\n    at Object.run (/snapshot/repo/build/node_modules/@aws-amplify/amplify-provider-awscloudformation/lib/push-resources.js:362:11)\r\n    at async /snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/extensions/amplify-helpers/push-resources.js:137:16\r\n    at async Promise.all (index 0)\r\n    at async providersPush (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/extensions/amplify-helpers/push-resources.js:133:5)\r\n    at async AmplifyToolkit.pushResources (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/extensions/amplify-helpers/push-resources.js:107:13)\r\n    at async Object.executeAmplifyCommand (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/index.js:194:9)\r\n    at async executePluginModuleCommand (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/execution-manager.js:139:5)\r\n    at async executeCommand (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/execution-manager.js:37:9)\r\n    at async Object.run (/snapshot/repo/build/node_modules/@aws-amplify/cli-internal/lib/index.js:121:5)\r\n\r\nResource is not in the state stackUpdateComplete\r\n"]
[214.861,"o","ResourceNotReady: Resource is not in the state stackUpdateComplete\r\n    at constructor.setError (/snapshot/repo/build/node_modules/aws-sdk/lib/resource_waiter.js:182:47)\r\n    at Request.CHECK_ACCEPTORS (/snapshot/repo/build/node_modules/aws-sdk/lib/resource_waiter.js:44:12)\r\n    at Request.callListeners (/snapshot/repo/build/node_modules/aws-sdk/lib/sequential_executor.js:106:20)\r\n    at Request.emit (/snapshot/repo/build/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\r\n    at Request.emit (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:686:14)\r\n    at Request.transition (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:22:10)\r\n    at AcceptorStateMachine.runTo (/snapshot/repo/build/node_modules/aws-sdk/lib/state_machine.js:14:12)\r\n    at /snapshot/repo/build/node_modules/aws-sdk/lib/state_machine.js:26:10\r\n    at Request.<anonymous> (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:38:9)\r\n    at Request.<anonymous> (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:688:12)\r\n    at Request.callListeners (/snapshot/repo/build/node_modules/aws-sdk/lib/sequential_executor.js:116:18)\r\n    at Request.emit (/snapshot/repo/build/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\r\n    at Request.emit (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:686:14)\r\n    at Request.transition (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:22:10)\r\n    at AcceptorStateMachine.runTo (/snapshot/repo/build/node_modules/aws-sdk/lib/state_machine.js:14:12)\r\n    at /snapshot/repo/build/node_modules/aws-sdk/lib/state_machine.js:26:10\r\n    at Request.<anonymous> (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:38:9)\r\n    at Request.<anonymous> (/snapshot/repo/build/node_modules/aws-sdk/lib/request.js:688:12)\r\n    at Request.callListeners (/snapshot/repo/build/node_modules/aws-sdk/lib/sequential_executor.js:116:18)\r\n    at callNextListener (/snapshot/repo/build/node_modules/aws-sdk/lib/sequential_executor.js:96:12)\r\n    at IncomingMessage.onEnd (/snapshot/repo/build/node_modules/aws-sdk/lib/event_listeners.js:417:13)\r\n    at IncomingMessage.emit (node:events:525:35)\r\n    at IncomingMessage.emit (node:domain:489:12)\r\n    at endReadableNT (node:internal/streams/readable:1359:12)"]
[214.861,"o","\r\n    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)\r\n"]
```

**Review by Amplifiyer — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

File: packages/amplify-migration-tests/package.json [hunk 6]
```diff
@@ -21,6 +21,7 @@
     "migration_v4.28.2_nonmultienv_layers": "yarn setup-profile 4.28.2 && jest src/__tests__/migration_tests/lambda-layer-migration/layer-migration.test.ts --verbose --config=jest.config.js",
     "migration_v4.52.0_multienv_layers": "yarn setup-profile 4.52.0 && jest src/__tests__/migration_tests/lambda-layer-migration/layer-migration.test.ts --verbose --config=jest.config.js",
     "migration_v10.5.1": "yarn setup-profile 10.5.1 && jest --verbose --config=jest.config.js",
+    "migration_v12.0.3": "yarn setup-profile 12.0.3 && jest --verbose --config=jest.config.js",
```
[the remark this answers is not in the document]
Comment (sobolk) [reply packages/amplify-migration-tests/package.json:24]: Current latest is soon to be replaced by `12.2.0-rc.a3b139fb36.0` . I'd refrain from jumping on moving target.
For this migration purposes `12.0.3` is as good as `latest`
The diffs (https://github.com/aws-amplify/amplify-cli/compare/v12.0.3...v12.1.1 and https://github.com/aws-amplify/amplify-cli/compare/v12.0.3...release_rc/a3b139fb36) don't reveal significant changes (both show 6 changed files - changelog, versions in package.json, test changes due to yarn migration and the `userPoolConsoleUrl` change - which isn't significant to this PR.).

If we ship this PR before batch 2 arrives then I'll be inclined to add one more v12 suite to cover version in-between.

**Pull request merged**